### PR TITLE
ucm2: Add intial Amlogic GXL support

### PR DESCRIPTION
**Pull request by jeromebrunet**
This adds the support for the GXL-P241 and the LIBRETECH-CC.
I have proposed the related ucm-validator PR: https://github.com/alsa-project/alsa-tests/pull/23

I was not successful with txt check:
```
!!Sysfs card info
!!---------------

Traceback (most recent call last):
  File "/home/jbrunet/work/generic/sound/alsa-tests/python/ucm-validator/./ucm.py", line 238, in <module>
    main(sys.argv)
  File "/home/jbrunet/work/generic/sound/alsa-tests/python/ucm-validator/./ucm.py", line 233, in main
    r = globals()[cmd](*argv[2:])
  File "/home/jbrunet/work/generic/sound/alsa-tests/python/ucm-validator/./ucm.py", line 185, in do_configs
    errors, warnings = config_walk(alsainfo_path)
  File "/home/jbrunet/work/generic/sound/alsa-tests/python/ucm-validator/./ucm.py", line 129, in config_walk
    errors1, warnings1 = config_walk(path2)
  File "/home/jbrunet/work/generic/sound/alsa-tests/python/ucm-validator/./ucm.py", line 142, in config_walk
    info.load(path2)
  File "/home/jbrunet/work/generic/sound/alsa-tests/python/ucm-validator/../lib/alsainfo.py", line 314, in load
    raise AlsaInfoError("%s: unknown section %s" % (filename, repr(name)))
alsainfo.AlsaInfoError: configs/Meson/LIBRETECH-CC.txt: unknown section 'Sysfs card info'
```

I also check with the dummy json:
```
*** configuration 'configs/Meson/GXL-P241.json' @ AlsaJsonSoundcard(json='GXL-P241', id='GXL-P241', driver='gx-sound-card', name='GXL-P241', longname='GXL-P241') ***
Device file 'conf.d/gx-sound-card/GXL-P241.conf'
Verb 'HiFi', file 'Meson/p241/p241-HiFi.conf'
Device 'Line'
Device 'HDMI'
ERR: conf.d/gx-sound-card/GXL-P241.conf: Verb=HiFi Device=Line PlaybackPCM defined in SectionVerb
ERR:   AlsaJsonSoundcard(json='GXL-P241', id='GXL-P241', driver='gx-sound-card', name='GXL-P241', longname='GXL-P241')
ERR: total errors: 1
```

I'm not quite sure where to put the `PlaybackPCM`, if not in `Verb`.
The `Device ` section in not a good fit, as far as I understand.
On these devices, we have a single PCM interface. HDMI and Line Devices may be enabled on it (possibly simultaneously).

Any suggestion ? 

Looking at pulseaudio, I'm quite happy with the result:
```
$ pactl list sinks
Sink #0
	State: SUSPENDED
	Name: alsa_output.platform-sound.HiFi__hw_LIBRETECHCC_0__sink
	Description: Built-in Audio HDMI + Analog Lineout
	Driver: module-alsa-card.c
	Sample Specification: s16le 2ch 44100Hz
	Channel Map: front-left,front-right
	Owner Module: 6
	Mute: no
	Volume: front-left: 65536 / 100% / 0.00 dB,   front-right: 65536 / 100% / 0.00 dB
	        balance 0.00
	Base Volume: 65536 / 100% / 0.00 dB
	Monitor Source: alsa_output.platform-sound.HiFi__hw_LIBRETECHCC_0__sink.monitor
	Latency: 0 usec, configured 0 usec
	Flags: HARDWARE DECIBEL_VOLUME LATENCY 
	Properties:
		alsa.resolution_bits = "16"
		device.api = "alsa"
		device.class = "sound"
		alsa.class = "generic"
		alsa.subclass = "generic-mix"
		alsa.name = ""
		alsa.id = "fe.dai-link-0 (*)"
		alsa.subdevice = "0"
		alsa.subdevice_name = "subdevice #0"
		alsa.device = "0"
		alsa.card = "0"
		alsa.card_name = "LIBRETECH-CC"
		alsa.long_card_name = "LIBRETECH-CC"
		alsa.driver_name = "snd_soc_meson_gx_sound_card"
		device.bus_path = "platform-sound"
		sysfs.path = "/devices/platform/sound/sound/card0"
		device.form_factor = "internal"
		device.string = "_ucm0001.hw:LIBRETECHCC,0"
		device.buffering.buffer_size = "352768"
		device.buffering.fragment_size = "176384"
		device.access_mode = "mmap+timer"
		device.profile.name = "HiFi: hw:LIBRETECHCC,0: sink"
		device.profile.description = "HDMI + Analog Lineout"
		alsa.mixer_device = "_ucm0001.hw:LIBRETECHCC"
		device.description = "Built-in Audio HDMI + Analog Lineout"
		module-udev-detect.discovered = "1"
		device.icon_name = "audio-card"
	Ports:
		[Out] HDMI: HDMI (type: HDMI, priority: 100, availability unknown)
		[Out] Line: Analog Lineout (type: Line, priority: 100, availability unknown)
	Active Port: [Out] HDMI
	Formats:
		pcm
```

### CODE REVIEW

**Comment by jeromebrunet** [comment]
Hello,
What can I do to progress on this ?

Thanks

File: ucm2/Meson/p241/p241-HiFi.conf [hunk 1]
```diff
@@ -0,0 +1,38 @@
+SectionVerb {
+	Value {
+		PlaybackPriority 100
+		PlaybackPCM "hw:${CardId},0"
```
Comment (perexg) [inline ucm2/Meson/p241/p241-HiFi.conf:4]: Move this (both) values to the section device configuration. Also, the priorities should be different (probably analog should have higher priority in this case - missing jack / link detection).

File: ucm2/Meson/p241/p241.conf [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Use case for the p241 Amlogic s805x reference design
```
Comment (perexg) [inline ucm2/Meson/p241/p241.conf:1]: What is 'Meson' as the directory name? The top level directory should be a SoC maker name.

**Review by perexg — CHANGES_REQUESTED** [review]
Thank you for your contribution. Please, try to address my comments.

File: ucm2/Meson/p241/p241-HiFi.conf (hunk 1 above)
Comment (jeromebrunet) [reply ucm2/Meson/p241/p241-HiFi.conf:4]: Well the pcm is not really linked the output device as I was explaining above.

The card behind this is a DPCM one. The PCMs are the frontends.
The output devices such as HDMI and Analog are the backend.
So in theory, the output device could be attached to different PCMs so I'm not sure it is correct to define the PCM there. 

There is a single PCM in this example - but I have another SoC getting ready where I have several generic PCMs and several outputs. The idea is too use 1 PCM per use-case (Hifi, Voice,) and route the PCM to appropriate output depending on the current use-case.

How would you handle this with UCM if verb is not appropriate ? 

---
For the priority, on these boards, the HDMI is used mostly. If I need to provide a different, I would suggest to put that higher.
There no jack detect on analog, indeed.
Ideally, I'd like the default configuration (priority?) to enable both outputs on the PCM. The HW supports it.

File: ucm2/Meson/p241/p241.conf (hunk 2 above)
Comment (jeromebrunet) [reply ucm2/Meson/p241/p241.conf:1]: Meson is name commonly used for Amlogic in the kernel - Like Sunxi for Allwinner
I'll change it to Amlogic

File: ucm2/Meson/p241/p241-HiFi.conf (hunk 1 above)
Comment (perexg) [reply ucm2/Meson/p241/p241-HiFi.conf:4]: The PA or PW use the PCM device name as the sink (thus if those names are equal - one PCM sink is created). Your configuration is also ok, but I would prefer to specify the PCM devices in the SectionDevice values.

PA or PW selects the default device by priority (and if equal - by the device order). If you like to enable both - you may create third conflicting device with appropriate control settings (enable both paths) with highest priority, but I would prefer to keep it separate and let PA or PW to handle it depending on the user selection. If the devices are not marked as conflicting, the applications may use both simultaneously.

> For the priority, on these boards, the HDMI is used mostly. If I need to provide a different, I would suggest to put that higher.

As you like.

Comment (perexg) [reply ucm2/Meson/p241/p241-HiFi.conf:4]: The priority is connected to the PCM device + mixer / path settings. If this combination is in a clash with another UCM device (PCM device + mixer / path settings), it should be marked as conflicting. The dynamic PCM routing in the driver should be described using a model which fits for the users - it usually means the static allocation of PCM devices for a purpose.

Note that only one verb can be active at a time, so it's something like a profile in pulseaudio.

**Comment by perexg** [comment]
Stalled.

**Pull request closed without merging**